src/Illuminate/Database/Concerns/ManagesTransactions.php

```diff
@@ -81,6 +93,35 @@ protected function handleTransactionException($e, $currentAttempt, $maxAttempts)
         throw $e;
     }
 
+    /**
+     * Handle an exception encountered when committing a transaction.
+     *
+     * @param  \Exception  $e
+     * @param  int  $currentAttempt
+     * @param  int  $maxAttempts
+     * @return void
+     *
+     * @throws \Exception
+     */
+    protected function handleCommitTransactionException($e, $currentAttempt, $maxAttempts)
+    {
+        // If the COMMIT fails, the transaction is automatically rolled back.
+        $this->transactions--;
+
+        // If we encountered a serialization failure, try again if we are not out of attempts.
+        // SQLSTATE 40001 is designated as a serialization failure.
```

Comment:
Uh, `SQLSTATE 40001` is database specific .. no? I don't think this is supposed to be in this "shared" handling code.

Why is only this particular error generating this problem and not others?

Reply:
Actually, no. SQLSTATE error codes are part of the SQL standard. PostgreSQL, MySQL, SQL Server, Oracle, and IBM databases all use SQLSTATE 40001 to report serialization failures.

Any exception occurring during commit would cause the rollback problem. We do need special handling for serialization failures, though, because that is a case where (just like a deadlock) retrying the transaction is the correct course of action.

Reply:
#TIL , thanks